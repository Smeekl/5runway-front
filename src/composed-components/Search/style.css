.search {
  padding: 62px 35px 0 35px;
}

.search__header {
  font-weight: 700;
  font-size: 24px;
}

.search__field {
  margin-bottom: 32px;
}

.search__suphistory {
  display: flex;
  justify-content: space-between;
  align-items: center;
}

.suphistory {
  margin-bottom: 16px;
}

.suphistory__tile {
  font-weight: 300;
  font-size: 12px;
  color: var(--color-secondary-gray);
}

.suphistory__link {
  cursor: pointer;
  font-weight: 300;
  font-size: 12px;
  color: var(--color-primary-blue);
  text-decoration-line: underline;
  text-underline-offset: 2px;
}

.history {
  margin-bottom: 40px;
}

.history__item {
  font-weight: 600;
  font-size: 16px;
  margin-bottom: 13px;
}

.history__item:last-child {
  margin-bottom: 0;
}

.search__show {
  background-color: var(--color-primary-blue);
  width: 343px;
  height: 48px;
  border-radius: 15px;
}

.search__show > span {
  color: var(--color-secondary-white)fff;
}

.search__show:hover,
.search__show:focus {
  background-color: var(--color-primary);
  border: none;
}
