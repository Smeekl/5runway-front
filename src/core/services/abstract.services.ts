import { AxiosInstance } from 'axios';
import { StorageService } from './storage-services/storage.services';
import { TTokens } from './types';

export class AbstractApiClient {
  api: AxiosInstance;
  storageService: StorageService;

  constructor(api: AxiosInstance, storageService: StorageService) {
    this.api = api;
    this.storageService = storageService;
  }

  setTokens({ accessToken, refreshToken }: TTokens) {
    this.storageService.set('accessToken', accessToken);
    this.storageService.set('refreshToken', refreshToken);
  }
}
