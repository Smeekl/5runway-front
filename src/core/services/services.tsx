import React from 'react';
import { IServices } from '@serviceTypes';

const { createContext, useContext } = React;
const ServicesContext = createContext({} as IServices);

type TServiceProviderProps = {
  services: IServices;
  children: React.ReactNode;
};

export const ServiceProvider = ({
  services,
  children,
}: TServiceProviderProps) => (
  <ServicesContext.Provider value={services}>
    {children}
  </ServicesContext.Provider>
);

export const useService = () => useContext(ServicesContext);
