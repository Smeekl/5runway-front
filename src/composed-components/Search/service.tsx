import React, { useState, useEffect } from 'react';
import { useService } from '@services/services';
import { TSearchProvider, TSearchValue } from '@composed-components/types';

const { createContext, useContext } = React;
const SearchContext = createContext({} as TSearchValue);

export const SearchProvider = ({ children }: TSearchProvider) => {
  const { searchService } = useService();

  const [history, setHistory] = useState<Array<string>>([]);
  const [inputValue, setInputValue] = useState('');

  useEffect(() => {
    setHistory(searchService.getSearchedItems() || []);
  }, []);

  function setInHistory() {
    if (!inputValue.trim()) return;

    searchService.searchItem(inputValue);
    setHistory((prevHistory) => [inputValue, ...prevHistory.slice(0, 9)]);
    setInputValue('');
  }

  function clearHistory() {
    searchService.clearSearchedItems();
    setHistory([]);
  }

  function handleSetHistory() {
    if (history.length <= 10) {
      return;
    }

    setHistory([...history.slice(0, 10)]);
  }

  return (
    <SearchContext.Provider
      value={{
        setInHistory,
        clearHistory,
        history,
        inputValue,
        setInputValue,
        handleSetHistory,
      }}
    >
      {children}
    </SearchContext.Provider>
  );
};

export const useSearch = () => useContext(SearchContext);
