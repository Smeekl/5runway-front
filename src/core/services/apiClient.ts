import Axios, { AxiosInstance } from 'axios';
import { StorageService } from './storage-services/storage.services';

export function CreateApiClient(
  baseURL: string,
  storageService: StorageService
): AxiosInstance {
  const api = Axios.create({ baseURL });

  let retry = false;

  api.interceptors.request.use(
    (config) => {
      if (!config.headers) {
        config.headers = {};
      }

      config.headers.Authorization = `Bearer ${storageService.get(
        'accessToken'
      )}`;

      return config;
    },
    (error) => Promise.reject(error)
  );
  api.interceptors.response.use(
    (res) => res,
    async (err) => {
      if (err.response.status === 401 && !retry) {
        retry = true;

        const refreshToken = storageService.get('refreshToken');

        if (!refreshToken) {
          return;
        }

        try {
          const { data } = await api.post('/auth/refresh', { refreshToken });

          storageService.set('accessToken', data.accessToken);
          storageService.set('refreshToken', data.refreshToken);
        } catch (error) {
          throw new Error('Token expired');
        }
        return;
      }

      retry = false;

      return Promise.reject(err);
    }
  );

  return api;
}
