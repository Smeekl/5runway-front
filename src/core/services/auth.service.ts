import { AbstractApiClient } from './abstract.services';
import { TTokens } from './types';

export class AuthService extends AbstractApiClient {
  googleAuth = async (google_token: string) => {
    const { data } = await this.api.get<TTokens>(`/auth/google/verification?token=${google_token}`);

    this.setTokens(data);
  }
}
