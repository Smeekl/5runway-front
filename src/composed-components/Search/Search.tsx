import { useEffect } from 'react';
import Button from '@components/Button/Button';
import SearchInput from '@components/SearchInput/SearchInput';
import { useSearch } from './service';
import './style.css';

const Search = () => {
  const {
    history,
    clearHistory,
    setInHistory,
    setInputValue,
    inputValue,
    handleSetHistory,
  } = useSearch();

  useEffect(() => {
    handleSetHistory();
  });

  return (
    <div className="search">
      <h1 className="search__header">Search</h1>
      <div className="search__field">
        <SearchInput
          value={inputValue}
          onChange={({ target }) => setInputValue(target.value)}
        />
      </div>
      <div className="search__suphistory suphistory">
        <span className="suphistory__tile">History</span>
        <span className="suphistory__link" onClick={clearHistory}>
          CLEAR
        </span>
      </div>
      <ul className="search__history history">
        {history.length ? (
          history.map((item, idx) => (
            <li className="history__item" key={idx}>
              {item}
            </li>
          ))
        ) : (
          <li>No history</li>
        )}
      </ul>
      <Button className="search__show" onClick={setInHistory}>
        Show results
      </Button>
    </div>
  );
};

export default Search;
