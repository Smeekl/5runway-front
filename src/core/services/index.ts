import { CreateApiClient } from './apiClient';
import { StorageService } from './storage-services/storage.services';
import { UserService } from './user.service';
import { SearchService } from './search.service';
import { AuthService } from './auth.service';

export function createServices() {
  const storageService = new StorageService(window.localStorage);
  const api = CreateApiClient(process.env.REACT_APP_BASEURL, storageService);
  const searchService = new SearchService(api, storageService);
  const userService = new UserService(api, storageService);
  const authService = new AuthService(api, storageService);

  return {
    storageService,
    searchService,
    userService,
    authService,
  };
}
