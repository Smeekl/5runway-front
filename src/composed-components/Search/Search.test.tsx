import { MemoryRouter } from 'react-router-dom';
import { render, waitFor } from '@testing-library/react';
import { ServiceProvider } from '@services/services';
import { createServices } from '@services/index';
import { SearchProvider } from './service';
import Search from './Search';

describe('RestorePassword UI tests', () => {
  test('should render correctly', () => {
    const services = createServices();

    const { container } = render(
      <MemoryRouter>
        <ServiceProvider services={services}>
          <SearchProvider>
            <Search />
          </SearchProvider>
        </ServiceProvider>
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  test('should be in the document', () => {
    const services = createServices();

    const { container } = render(
      <MemoryRouter>
        <ServiceProvider services={services}>
          <SearchProvider>
            <Search />
          </SearchProvider>
        </ServiceProvider>
      </MemoryRouter>
    );
    expect(container).toBeInTheDocument();
  });

  test("component shouldn't render more that 10 elements", () => {
    const services = createServices();

    const arr = [];

    for (let i = 0; i < 50; i++) {
      arr.push('test');
    }

    global.localStorage.setItem('searchedItem', JSON.stringify(arr));

    const { getAllByText } = render(
      <MemoryRouter>
        <ServiceProvider services={services}>
          <SearchProvider>
            <Search />
          </SearchProvider>
        </ServiceProvider>
      </MemoryRouter>
    );

    waitFor(() => expect(getAllByText('test')).toHaveLength(10));
  });

  test('should clear history', () => {
    const services = createServices();

    const arr = [];

    for (let i = 0; i < 50; i++) {
      arr.push('test');
    }

    global.localStorage.setItem('searchedItem', JSON.stringify(arr));

    const { getAllByText, getByText } = render(
      <MemoryRouter>
        <ServiceProvider services={services}>
          <SearchProvider>
            <Search />
          </SearchProvider>
        </ServiceProvider>
      </MemoryRouter>
    );

    waitFor(() => {
      getByText('CLEAR').click();
      expect(global.localStorage.getItem('searchedItem') as string).toBeNull();
      expect(getAllByText('test')).toHaveLength(0);
    });
  });
});
