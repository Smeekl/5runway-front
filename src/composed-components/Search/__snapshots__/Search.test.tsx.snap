// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestorePassword UI tests should render correctly 1`] = `
<div>
  <div
    class="search"
  >
    <h1
      class="search__header"
    >
      Search
    </h1>
    <div
      class="search__field"
    >
      <span
        class="ant-input-affix-wrapper search__field"
      >
        <input
          class="ant-input"
          type="text"
          value=""
        />
        <span
          class="ant-input-suffix"
        >
          <reactcomponent
            class="search__icon"
          />
        </span>
      </span>
    </div>
    <div
      class="search__suphistory suphistory"
    >
      <span
        class="suphistory__tile"
      >
        History
      </span>
      <span
        class="suphistory__link"
      >
        CLEAR
      </span>
    </div>
    <ul
      class="search__history history"
    >
      <li>
        No history
      </li>
    </ul>
    <button
      class="ant-btn ant-btn-default search__show"
      type="button"
    >
      <span>
        Show results
      </span>
    </button>
  </div>
</div>
`;
